rocksdb_replicator/tests/rocksdb_replicator_test.cpp

```diff
@@ -348,6 +361,126 @@ TEST(RocksDBReplicatorTest, 1_master_2_slaves_chain) {
   EXPECT_EQ(db_slave_2->GetLatestSequenceNumber(), 2 * n_keys);
 }
 
+TEST(RocksDBReplicatorTest, 1_master_1_slave_upstream_itself) {
+  FLAGS_replicator_max_server_wait_time_ms = 100;
+  FLAGS_replicator_client_server_timeout_difference_ms = 100;
+  FLAGS_reset_upstream_on_empty_updates_from_non_leader = true;
+  FLAGS_replicator_max_consecutive_no_updates_before_upstream_reset = 1;
+
+  int16_t master_port = 9092;
+  int16_t slave_port = 9093;
+  Host master(master_port);
+  Host slave(slave_port);
+
+  auto db_master = cleanAndOpenDB("/tmp/db_master");
+  auto db_slave = cleanAndOpenDB("/tmp/db_slave");
+
+  RocksDBReplicator::ReplicatedDB* replicated_db_master;
+  RocksDBReplicator::ReplicatedDB* replicated_db_slave;
+
+  EXPECT_EQ(master.replicator_->addDB("shard1", db_master, ReplicaRole::LEADER, folly::SocketAddress(),
+                              &replicated_db_master),
+            ReturnCode::OK);
+
+  // follower setting itself as the upstream, hence it won't receive updates from leader,
+  // unless the upstream is reset to be the leader.
+  SocketAddress addr_slave("127.0.0.1", slave_port);
+  EXPECT_EQ(slave.replicator_->addDB("shard1", db_slave, ReplicaRole::FOLLOWER,
+                                     addr_slave, &replicated_db_slave),
+            ReturnCode::OK);
+
+  EXPECT_EQ(db_master->GetLatestSequenceNumber(), 0);
+  EXPECT_EQ(db_slave->GetLatestSequenceNumber(), 0);
+
+  WriteOptions options;
+  uint32_t n_keys = 100;
+  for (uint32_t i = 0; i < n_keys; ++i) {
+    WriteBatch updates;
+    auto str = to_string(i);
+    updates.Put(str + "key", str + "value");
+    updates.Put(str + "key2", str + "value2");
+    EXPECT_EQ(master.replicator_->write("shard1", options, &updates),
+              ReturnCode::OK);
+    EXPECT_EQ(db_master->GetLatestSequenceNumber(), i * 2 + 2);
+  }
+
+  // expect follower reset upstream to be triggered
```

Comment:
when the follower trigger the `resetUpstream`, it seem like the `resetUpstream` will request zookeeper?
Now, at this unit test, there is no zk setup ,right? Then, what is the side effect of that `resetUpstream` call?

Reply:
yeah, it'll just be a no-op: https://github.com/pinterest/rocksplicator/blob/bec3645b49c996e3321faac5ab52b9fe04df1a51/rocksdb_replicator/replicated_db.cpp#L231, I have a comment/test below to verify the `resetUpstreamAttempts_` is incremented, but no sequence number change (because reset won't succeed due to this)